### PR TITLE
ipq95xx: Add support for IPQ9574 RDP433

### PR DESCRIPTION
**Pull request by quic-varada**
Hi,

This pull request is for adding support to the QCOM IPQ9574 SoC.
Kindly review and provide feedback.

SoC	: QCOM IPQ9574
RAM	: 2GB DDR4
Flash	: eMMC 8GB
WiFi	: 1 x 2.4GHz
	  1 x 5GHz
	  1 x 6GHz

Thanks
Varada

### CODE REVIEW

**Comment by xback** [comment]
@quic-varada
_ipq95xx: Enable RPM based regultors_ --> regulators

**Comment by mrkiko** [comment]
Mark to follow.

**Comment by quic-varada** [comment]
> @quic-varada _ipq95xx: Enable RPM based regultors_ --> regulators
> 
> Also, get rid of the merge commit:
> 
> * pull + rebase
> * force-push

Thanks for the feedback. Have updated and pushed once again.
Please take a look.

**Comment by 981213** [comment]
Hi! This PR looks good to me so far.
But as we discussed before on mail, a serial-only target isn't really usable for OpenWrt. It should at least get working ethernet before it can be merged.
Don't be hesitant to keep working on it and push further changes here for reviewing :)

**Comment by rmandrad** [comment]
> Hi! This PR looks good to me so far. But as we discussed before on mail, a serial-only target isn't really usable for OpenWrt. It should at least get working ethernet before it can be merged. Don't be hesitant to keep working on it and push further changes here for reviewing :)

what is the device here ? a target is fine but where is this has been tested and proven?

**Review by ynezz — CHANGES_REQUESTED** [review]
@quic-varada thanks, but is there any reason, why you didn't already incorporated the review feedback provided [on the patch](https://patchwork.ozlabs.org/project/openwrt/patch/20231207095923.3913903-1-quic_varada@quicinc.com/) you've sent to the mailing list?

I see namely following most glaring points:

1. [Kernel patches - please sort the patches by prefix such as](https://patchwork.ozlabs.org/project/openwrt/patch/20231207095923.3913903-1-quic_varada@quicinc.com/#3229858):
    * 0xx are backports (Kernel version from which they are backported must be marked as well)
       * `ls -al target/linux/qualcommax/patches-6.1/*v6.*` for prior examples  
    * 1xx are pending
    * 9xx are usually hacks/stuff that currently cannot be upstreamed

2. [target and subtarget name](https://patchwork.ozlabs.org/project/openwrt/patch/20231207095923.3913903-1-quic_varada@quicinc.com/#3230401)
   * IIUC then the current preference is to use `qualcommax/ipq95xx`, we can always rename/split it later, so no worries, its not something set into the stone.

In other words, those are mandatory.

**Comment by ynezz** [comment]
> a serial-only target isn't really usable for OpenWrt
> It should at least get working ethernet before it can be merged.

I would say, that its not strictly mandatory, see such prior examples 29c6ddd99ee and 3c1f6e358d. I for example find this smaller diffs much easier for handling/reasoning about for both sides (contributor/reviewer). Looking at the diff, its mostly upstream backports right now.

And there was a promise made: [Yes, there is plan to post those functionalities in the coming weeks.](https://patchwork.ozlabs.org/project/openwrt/patch/20231207095923.3913903-1-quic_varada@quicinc.com/#3230401) 

> what is the device here ?

IIUC then its a `ap-al02.1-c4` board targeting developers, so QCA's development kit.

> a target is fine but where is this has been tested and proven?

Yep, I'll test this.

File: target/linux/ipq95xx/base-files/etc/inittab [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+# purpose with or without fee is hereby granted, provided that the above
+# copyright notice and this permission notice appear in all copies.
+#
+# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
```
[the remark this answers is not in the document]
Comment (quic-varada) [reply target/linux/ipq95xx/base-files/etc/inittab:14]: My mistake.
Will make the corrections and post a new one.

Thanks

**Comment by huangya90** [comment]
Hi, does ipq95xx still use "switch_mac_mode" to configure interface mode? If yes, why I can not see them in the dts?

**Comment by Yewenzhou1994** [comment]
您好，请问可以适配一下 TP-Link TL-7TR13090 吗？这个机器能使用320Mhz

**Comment by AEGEGE** [comment]
Such a good project hasn't been updated for a long time, what's going on?

**Comment by tr0llex** [comment]
@ynezz @quic-varada Hello there. The community of Xiaomi Router 7000 (Xiaomi Router BE7000) owners, represented by the 4PDA forum members, is really looking forward to adding support for this SoC. Can we help in any way to get this Pull Request moving forward?

**Comment by zinky81** [comment]
Also new routers are appearing with IPQ9574. The latest one is Xiaomi 10G (BE10000). This should be very interesting router as it is successor of very popular Xiaomi AX9000, but with support of WIFI-7. This pull request should have a much more attention as device list is growing rapidly.

**Comment by zmashine** [comment]
**[@ynezz](https://github.com/ynezz)** [@quic-varada](https://github.com/quic-varada)
 
You are doing a very good thing. People crave more freedom and hope for you very much. Bring the case to the end. Thank you very much.

**Comment by zmashine** [comment]
@ynezz @quic-varada
 
You are doing a very good thing. People crave more freedom and hope for you very much. Bring the case to the end. Thank you very much.

**Comment by frogost** [comment]
I have a Xiaomi Router 7000 (Xiaomi Router BE7000) and I am really looking forward to the implementation of this project

**Comment by quic-varada** [comment]
Hi All,

the above comments are addressed in updated PR. Kindly help to review.

Thanks,
Varada

**Comment by 981213** [comment]
Hello! It's been half a year since the promised ethernet support. Any updates on that?

**Comment by BotMolotovich** [comment]
Hi! watching this pr for monthes, will so much appreciate any updates here!

**Comment by ArtMvs** [comment]
I have been following since February and every month I am disappointed, when will you stop our suffering? Why does your direct competitor mtk provide all the sources and moreover helps to develop openwrt?

Xiaomi be7000

**Comment by isikerkan** [comment]
> I have been following since February and every month I am disappointed, when will you stop our suffering? Why does your direct competitor mtk provide all the sources and moreover helps to develop openwrt?
> 
> Xiaomi be7000

Problem here is not the Devs @ArtMvs . Qualcom Employees coded the Drivers and submitted the fixes, but no reviewer checked them out yet. @ynezz Could you check the updated PR out?

We are eagerly waiting to flash the BE10000 to OpenWRT to have a affordable 10 Gbit/s SPF Router with a customizable OS.

**Comment by 981213** [comment]
@isikerkan Please note that this PR does not bring any networking support, meaning it's far from being usable on a real router.
The currently submitted source code has been reviewed and acked a long time ago, since it contains mostly backported patches. I just don't feel like merging such a target where nothing other than the serial console works.

File: target/linux/qualcommax/patches-6.6/0409-Squashfs-Accept-mtd-names-for-mounting.patch [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
++#include <linux/syscalls.h>
++#include <linux/mtd/mtd.h>
++#include <linux/init_syscalls.h>
+ 
+ #include "squashfs_fs.h"
+ #include "squashfs_fs_sb.h"
+@@ -480,8 +484,44 @@ failed_mount:
+ 	return err;
+ }
+
```
Comment (981213) [inline target/linux/qualcommax/patches-6.6/0409-Squashfs-Accept-mtd-names-for-mounting.patch:38]: Drop this patch. This hack shouldn't be needed on mainline OpenWrt.

File: target/linux/qualcommax/patches-6.6/0411-fix-build-errors.patch [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+ 
+ static const struct qcom_nandc_props ipq9574_snandc_props = {
+--- a/drivers/mtd/mtd_blkdevs.c
++++ b/drivers/mtd/mtd_blkdevs.c
+@@ -481,7 +481,7 @@ static void blktrans_notify_add(struct m
+ {
+ 	struct mtd_blktrans_ops *tr;
+ 
+-	if (mtd->type == MTD_ABSENT || mtd->type == MTD_UBIVOLUME)
++	if (mtd->type == MTD_ABSENT)
```
Comment (981213) [inline target/linux/qualcommax/patches-6.6/0411-fix-build-errors.patch:46]: Drop this change to the mtd core unless you have a really good reason to do so. If it's absolutely needed, please split it into a separated patch and explain why in the mbox header of the new patch.

**Review by 981213 — CHANGES_REQUESTED** [review]

**Comment by AEGEGE** [comment]
It seems that QCOM IPQ9574 will have a long way to go to successfully run openwrt. I am disappointed.

**Comment by RightDecision2** [comment]
> It seems that QCOM IPQ9574 will have a long way to go to successfully run openwrt. I am disappointed.

Why you cant take needed files/drivers from Xiaomi BE7000?
He ships with OpenWRT 18.06_Snapshot and all works well with QCOM

**Comment by zinky81** [comment]
This topic about support on IPQ9574 is somehow strange, because there is already functional QWRT firmware for Xiaomi BE10000 and BE7000 which is possible to purchase somewhere on Chinese and Russian forums. I'm also running this firmware and I can confirm it's running on kernel version 5.4.164. I have also tried to compile my own OpenWRT because there is already listed IPQ9574 support but it was unsuccessful due to limitations in Linux kernel for BE (Wifi-7). So for sure it can work, but the question is why is not open for OPENWRT. Thanks!

**Comment by waitingsnow** [comment]
> This topic about support on IPQ9574 is somehow strange, because there is already functional QWRT firmware for Xiaomi BE10000 and BE7000 which is possible to purchase somewhere on Chinese and Russian forums. I'm also running this firmware and I can confirm it's running on kernel version 5.4.164. I have also tried to compile my own OpenWRT because there is already listed IPQ9574 support but it was unsuccessful due to limitations in Linux kernel for BE (Wifi-7). So for sure it can work, but the question is why is not open for OPENWRT. Thanks!

How can I get the firmware you mentioned? I need it very much, thank you

**Comment by AEGEGE** [comment]
> This topic about support on IPQ9574 is somehow strange, because there is already functional QWRT firmware for Xiaomi BE10000 and BE7000 which is possible to purchase somewhere on Chinese and Russian forums. I'm also running this firmware and I can confirm it's running on kernel version 5.4.164. I have also tried to compile my own OpenWRT because there is already listed IPQ9574 support but it was unsuccessful due to limitations in Linux kernel for BE (Wifi-7). So for sure it can work, but the question is why is not open for OPENWRT. Thanks!

Xiaomi BE10000 hardware is not fully functional. The TP-LINK TL-7TR-13090 BE13000 I have has basically full performance. I can't find any information about shell and openwrt. I don't know when I can get it.

**Comment by RightDecision2** [comment]
[Xiaomi 7000.zip](https://github.com/user-attachments/files/18160232/Xiaomi.7000.zip)
Please review these files from Xiaomi BE7000.
IPQ9574.bin -> /tmp/IPQ9574/caldata.bin
qcn9224.bin -> /tmp/qcn9224/caldata_3.bin

**Comment by ndoo** [comment]
More IPQ95xx targets that have retail availability and would greatly benefit from this PR as a starting point:

- [Alta Labs Route10](https://www.alta.inc/route10)
- Most TP-Link 802.11be Products (Deco BE95, Deco BE85, EB810v, HB810)

+1 to get this going, meanwhile will experiment locally with QSDK tree.

**Comment by A-SOM** [comment]
any updates? What can I do to help speed up implementation?

**Comment by dima424658** [comment]
> How can I get the firmware you mentioned? I need it very much, thank you

@waitingsnow , here is the firmware for be7000 - https://cdn.cnbj1.fds.api.mi-img.com/xiaoqiang/rom/rc06/miwifi_rc06_firmware_93757_1.1.16.bin
it can be downloaded from - https://www1.miwifi.com/miwifi_download.html

**Comment by A-SOM** [comment]
> I have not had time to give it a go, but in the meantime, I have requested and received the GPL archive for Alta Labs AP6:
> 
> https://dl.alta.inc/gpl/gpl-2.2f-zkdp29d9.tar.gz

Hello. You can check gpl archive too here for MBE70:
https://support.linksys.com/kb/article/316-en/

https://downloads.linksys.com/support/assets/gpl/MBE70WRT_v1.1.0.01.tgz

**Comment by PerezChilli** [comment]
Mee too join and up this thread. I'm hope support will be reached cause IPQ95XX becomes more and more popular platform...

**Comment by ynezz** [comment]
Thanks, but it was superseded by https://github.com/openwrt/openwrt/pull/17725

**Comment by peterwillcn** [comment]
any news?
